modules/tensor_mechanics/src/materials/PlasticTruss.C

```diff
@@ -0,0 +1,152 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PlasticTruss.h"
+#include "Function.h"
+#include "MooseException.h"
+#include "MathUtils.h"
+
+registerMooseObject("TensorMechanicsApp", PlasticTruss);
+
+template <>
+InputParameters
+validParams<PlasticTruss>()
+{
+  InputParameters params = validParams<LinearElasticTruss>();
+  params.addClassDescription("Computes the stress and strain in a truss element considering J2 "
+                             "plasticity with linear hardening.");
+  params.addRequiredParam<Real>(
+      "yield_stress", "Input data for yield stress after which plastic strain starts accumulating");
+  params.addParam<Real>("hardening_constant", 0.0, "Hardening slope");
+  params.addParam<FunctionName>("hardening_function",
+                                "True stress as a function of plastic strain");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-10, "Absolute convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for Newton iteration");
+  return params;
+}
+
+PlasticTruss::PlasticTruss(const InputParameters & parameters)
+  : LinearElasticTruss(parameters),
+    _yield_stress(getParam<Real>("yield_stress")), // Read from input file
+    _hardening_constant(getParam<Real>("hardening_constant")),
+    _hardening_function(isParamValid("hardening_function") ? &getFunction("hardening_function")
+                                                           : NULL),
+    _hardening_slope(0.0),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _total_stretch_old(getMaterialPropertyOld<Real>(_base_name + "total_stretch")),
+    _plastic_strain(declareProperty<Real>(_base_name + "plastic_stretch")),
+    _plastic_strain_old(getMaterialPropertyOld<Real>(_base_name + "plastic_stretch")),
+    _elastic_strain_old(getMaterialPropertyOld<Real>(_base_name + "elastic_stretch")),
+    _stress_old(getMaterialPropertyOld<Real>(_base_name + "axial_stress")),
+    _strain_increment(declareProperty<Real>(_base_name + "strain_increment_truss")),
+    _hardening_variable(declareProperty<Real>(_base_name + "hardening_variable_truss")),
+    _hardening_variable_old(getMaterialPropertyOld<Real>(_base_name + "hardening_variable_truss")),
+    _max_its(1000)
+{
+  if (!parameters.isParamSetByUser("hardening_constant") && !isParamValid("hardening_function"))
+    mooseError("Either hardening_constant or hardening_function must be defined");
+
+  if (parameters.isParamSetByUser("hardening_constant") && isParamValid("hardening_function"))
+    mooseError(
+        "Only the hardening_constant or only the hardening_function can be defined but not both");
+}
+
+void
+PlasticTruss::initQpStatefulProperties()
+{
+  LinearElasticTruss::initQpStatefulProperties();
+  _plastic_strain[_qp] = 0.0;
+  _hardening_variable[_qp] = 0.0;
```

Comment:
That's done in the base class `LinearElasticTruss`.

Reply:
I see.  Actually in TrussMaterial.  It actually shouldn't be done there because they're not stateful in that class.  We're only supposed to set values in `initQpStatefulProperties()` for properties that are stateful.  As part of this PR, could you remove `TrussMaterial::initQpStatefulProperties()`? None of the properties that are set there are stateful.